File: docs/_src/utils/UserSettingsDocs.cs
using Newtonsoft.Json.Linq;
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using Xarial.XToolkit.Services.Data;
using Xarial.XToolkit.Services.Data.Attributes;

namespace Utils.Docs
{
    public static class UserSettingsDocs
    {
        public static class V1
        {
            //--- v1
            [DataVersion("1.0.0", typeof(UserSettingsVersionTransformer))]
            public class UserSettings
            {
                public string Field1 { get; set; }
                public double Field2 { get; set; }
                public bool Field3 { get; set; }
            }
            //---
        }

        public static class V2
        {
            //--- v2
            [DataVersion("2.0.0", typeof(UserSettingsVersionTransformer))]
            public class UserSettings
            {
                public string TextField { get; set; }
                public double Field2 { get; set; }
                public bool Field3 { get; set; }
            }
            //---
        }

        //--- v3
        [DataVersion("3.0.0", typeof(UserSettingsVersionTransformer))]
        public class UserSettings
        {
            public string TextField { get; set; }
            public double DoubleField { get; set; }
            public bool BoolField { get; set; }
        }
        //---

        //--- transformer
        public class UserSettingsVersionTransformer : IVersionsTransformer
        {
            public IReadOnlyList<VersionTransform> Transforms { get; }

            public UserSettingsVersionTransformer()
            {
                Transforms = new VersionTransform[]
                {
                    new VersionTransform(new Version("1.0.0"), new Version("2.0.0"), t =>
                    {
                        var field1 = ((JObject)t).Property("Field1");
                        field1.Replace(new JProperty("TextField", field1.Value));
                        return t;
                    }),
                    new VersionTransform(new Version("2.0.0"), new Version("3.0.0"), t =>
                    {
                        var field2 = ((JObject)t).Property("Field2");
                        field2.Replace(new JProperty("DoubleField", field2.Value));

                        var field3 = ((JObject)t).Property("Field3");
                        field3.Replace(new JProperty("BoolField", field3.Value));

                        return t;
                    })
                };
            }
        }
        //---

        public static void StoreSettings() 
        {
            //--- store
            var svc = new NsJsonDataSerializer<UserSettings>();
            
            var userSetts = new UserSettings();

            svc.Save(userSetts,
                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), 
                "my-app-settings.json"));

            //---
        }

        public static void ReadSettings()
        {
            //--- read
            var svc = new NsJsonDataSerializer<UserSettings>();

            var userSetts = svc.Read(
                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
                "my-app-settings.json"));
            //---
        }
    }
}

File: docs/_src/utils/UserSettingsDocs.vb
Imports Newtonsoft.Json.Linq
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text
Imports Xarial.XToolkit.Services.Data
Imports Xarial.XToolkit.Services.Data.Attributes

Namespace Utils.Docs

    Module V1
        '--- v1
        <DataVersion("1.0.0", GetType(UserSettingsVersionTransformer))>
        Public Class UserSettings
            Public Property Field1 As String
            Public Property Field2 As Double
            Public Property Field3 As Boolean
        End Class
        '---
    End Module

    Module V2
        '--- v2
        <DataVersion("2.0.0", GetType(UserSettingsVersionTransformer))>
        Public Class UserSettings
            Public Property TextField As String
            Public Property Field2 As Double
            Public Property Field3 As Boolean
        End Class
        '---
    End Module

    Module UserSettingsDocs

        '--- v3
        <DataVersion("3.0.0", GetType(UserSettingsVersionTransformer))>
        Public Class UserSettings
            Public Property TextField As String
            Public Property DoubleField As Double
            Public Property BoolField As Boolean
        End Class
        '---

        '--- transformer
        Public Class UserSettingsVersionTransformer
            Implements IVersionsTransformer

            Public Sub New()
                Transforms = New VersionTransform() {
                    New VersionTransform(New Version("1.0.0"), New Version("2.0.0"),
                                         Function(o)
                                             Dim field1 = o.Property("Field1")
                                             field1.Replace(New JProperty("TextField", field1.Value))
                                             Return o
                                         End Function),
                    New VersionTransform(New Version("2.0.0"), New Version("3.0.0"),
                                         Function(o)
                                             Dim field2 = o.Property("Field2")
                                             field2.Replace(New JProperty("DoubleField", field2.Value))
                                             Dim field3 = o.Property("Field3")
                                             field3.Replace(New JProperty("BoolField", field3.Value))
                                             Return o
                                         End Function)}
            End Sub

            Public ReadOnly Property Transforms As IReadOnlyList(Of VersionTransform) Implements IVersionsTransformer.Transforms

        End Class
        '---

        Sub StoreSettings()
            '--- store
            Dim svc = New NsJsonDataSerializer(Of UserSettings)()
            Dim userSetts = New UserSettings()
            svc.Save(userSetts,
                              Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
                                           "my-app-settings.json"))
            '---
        End Sub

        Sub ReadSettings()
            '--- read
            Dim svc = New NsJsonDataSerializer(Of UserSettings)()
            Dim userSetts = svc.Read(
                Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData),
                             "my-app-settings.json"))
            '---
        End Sub
    End Module
End Namespace
